enterprise/trackedusers/users_reporter_test.go

```diff
@@ -0,0 +1,304 @@
+			UserIDHll:                &userIDHll,
+			AnonymousIDHLL:           &annIDHll,
+			IdentifiedAnonymousIDHLL: &identifiedAnnIDHll,
+		}
+	}
+)
+
+func TestUniqueUsersReporter(t *testing.T) {
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	t.Run("GenerateReportsFromJobs", func(t *testing.T) {
+		testCases := []struct {
+			name             string
+			jobs             []*jobsdb.JobT
+			sourceIDtoFilter map[string]bool
+			trackedUsers     map[string]map[string]int
+		}{
+			{
+				name: "happy case",
+				jobs: []*jobsdb.JobT{
+					sampleTestJob1,
+					sampleTestJob2,
+					sampleTestJob3,
+					sampleTestJob4,
+				},
+				trackedUsers: map[string]map[string]int{
+					"workspaceID": {
+						"sourceID": 2,
+					},
+				},
+			},
+			{
+				name: "happy case 2",
+				jobs: []*jobsdb.JobT{
+					sampleTestJob1,
+					sampleTestJob2,
+					sampleTestJob3,
+					sampleTestJob4,
+					prepareJob(sampleSourceID, "user", "ann", sampleWorkspaceID),
+					prepareJob(sampleSourceID, "user", "ann", sampleWorkspaceID2),
+				},
+				trackedUsers: map[string]map[string]int{
+					sampleWorkspaceID: {
+						sampleSourceID: 3,
+					},
+					sampleWorkspaceID2: {
+						sampleSourceID: 1,
+					},
+				},
+			},
+			{
+				name: "filter non event stream sources",
+				jobs: []*jobsdb.JobT{
+					sampleTestJob1,
+					sampleTestJob2,
+					sampleTestJob3,
+					sampleTestJob4,
+					prepareJob(sampleSourceID, "user", "ann", sampleWorkspaceID),
+					prepareJob(sampleSourceID, "user", "ann", sampleWorkspaceID2),
+					prepareJob(sampleSourceToFilter, uuid.NewString(), uuid.NewString(), sampleWorkspaceID2),
+				},
+				sourceIDtoFilter: map[string]bool{
+					sampleSourceToFilter: true,
+				},
+				trackedUsers: map[string]map[string]int{
+					sampleWorkspaceID: {
+						sampleSourceID: 3,
+					},
+					sampleWorkspaceID2: {
+						sampleSourceID: 1,
+					},
+				},
+			},
+			{
+				name: "no source id in job parameters",
+				jobs: []*jobsdb.JobT{
+					prepareJob("", "user", "ann", sampleWorkspaceID),
+					prepareJob("", "user", "ann", sampleWorkspaceID),
+				},
+				trackedUsers: map[string]map[string]int{},
+			},
+			{
+				name: "no event tracked in reports",
+				jobs: []*jobsdb.JobT{
+					prepareJob("source", "", "", sampleWorkspaceID),
+					prepareJob("", "user", "ann", sampleWorkspaceID),
+					prepareJob("", "user", "ann", sampleWorkspaceID),
+				},
+				trackedUsers: map[string]map[string]int{},
+			},
+			{
+				name: "no workspace id in job",
+				jobs: []*jobsdb.JobT{
+					prepareJob("source", "user", "ann", ""),
+					prepareJob("source", "user", "ann", ""),
+				},
+				trackedUsers: map[string]map[string]int{},
+			},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				collector := &UniqueUsersReporter{log: logger.NOP, hllSettings: &hllSettings}
+				reports := collector.GenerateReportsFromJobs(tc.jobs, tc.sourceIDtoFilter)
+
+				result := make(map[string]map[string]int)
```

Comment:
It would be preferable to explicitly define your result instead of generating it. It is only a few fields, it would make your test cases more readable and easy to reason about.

Also, it prevents the same bugs in both generation and code from being hidden.